csharp/src/iceboxnet/ServiceManager.cs

```diff
@@ -406,33 +356,36 @@ private void RemoveObserver(IServiceObserverPrx observer, Exception ex)
             {
                 if (_observers.Remove(observer))
                 {
-                    // CommunicatorDestroyedException may occur during shutdown. The observer notification has been sent,
+                    // CommunicatorDisposedException may occur during shutdown. The observer notification has been sent,
                     // but the communicator was destroyed before the reply was received. We do not log a message for this
                     // exception.
                     if (_traceServiceObserver >= 1 && !(ex is CommunicatorDisposedException))
                     {
                         _logger.Trace("IceBox.ServiceObserver",
-                                      $"Removed service observer {observer}\nafter catching {ex}");
+                                      $"removed service observer {observer}\nafter catching {ex}");
                     }
                 }
             }
         }
 
-        private void ServiceStarted(string service, IEnumerable<IServiceObserverPrx> observers)
+        private void ServiceStarted(string name)
```

Comment:
If ServiceStarted/ServiceStoped aren't called with the mutex locked anymore, I don't see how we prevent potential out-of-order observer notifications.

Reply:
I think the existing code only gives the impression that out-of-order notifications are not possible.

Since we don't wait for the response from the observers (which would be bad within a mutex lock), there is no guarantee an earlier call reaches the observer before a later call.

Reply:
Fixed.